tests/test_variable_routine_testing.py

```diff
@@ -0,0 +1,295 @@
+        self.rng = random.Random(1)
+        # Output class requires a file name, but we don't write to disk in these tests,
+        # so we give it a dummy file name.
+        self.out = Output(file_name="unused")
+        # Sex and race_ethnicity are irrelevant to this test but we need to choose an
+        # arbitrary value for the simulation to run.
+        self.sex = "female"
+        self.race_ethnicity = "black_non_hispanic"
+
+    def start(self):
+        self.choose_tests()
+
+        self.handle_disease_message(PersonDiseaseMessage.INIT)
+        self.handle_testing_message(PersonTestingMessage.INIT)
+        self.handle_treatment_message(PersonTreatmentMessage.INIT)
+
+        self.scheduler.add_event(
+            message="Conduct yearly actions",
+            delay=1,
+            handler=self.handle_yearly_actions,
+        )
+
+        # Fix lifespan at 100 for testing instead of calling self.start_life_timer()
+        self.expected_lifespan = 100
+        self.scheduler.add_event(
+            message=PersonDiseaseMessage.OTHER_DEATH,
+            handler=self.handle_disease_message,
+            delay=self.expected_lifespan,
+        )
+        self.out.add_expected_lifespan(
+            person_id=self.id,
+            time=self.expected_lifespan,
+        )
+
+        # Person.start has lesion delay functions here to add an event to the
+        # scheduler for the person's first lesion. We don't want any lesions for the
+        # test person, so that chunk is omitted here. Because the next lesion delay
+        # is computed when a lesion onset is handled, this results in the person
+        # never having a lesion.
+
+    def simulate(self):
+        """
+        Simplified version of the simulation loop used in crcsim.__main__.
+        Enables us to simulate one TestPerson at a time without running the
+        main simulation on a cohort of people.
+        """
+        while not self.scheduler.is_empty():
+            event = self.scheduler.consume_next_event()
+            if not event.enabled:
+                continue
+            if event.message == "end_simulation":
+                logging.debug("[scheduler] ending simulation \n")
+                break
+            handler = event.handler
+            handler(event.message)
+
+
+def test_testing_year_misalignment(params):
+    """
+    Asserts that misaligned testing years in routine_testing_year and a single test's
+    routine start or end raises an error.
+    """
+    with open("parameters.json", "r") as f:
+        params = json.load(f)
+        params["tests"]["Colonoscopy"]["routine_end"] = 85
+        with TemporaryDirectory() as tmp_dir:
+            tmp_path = Path(tmp_dir) / "parameters.json"
+            with open(tmp_path, "w") as f:
+                json.dump(params, f)
+                with pytest.raises(ValueError):
+                    load_params(tmp_path)
+
+
+def test_one_colonoscopy_equivalence(params, test_switching_scenarios):
```

Comment:
I think the way you set up these test-switching tests is reasonable, and if you'd prefer to keep it this way I won't hold up the PR. But I do observe that these three tests are essentially identical, so there's duplicated logic that could be eliminated if these were combined into a single test. I think you can do that by parameterizing one test with cases that look like:

```
@pytest.mark.parametrize(
    "case",
    [
        # switches to FIT at age 51, but they shouldn't get a FIT test until age 60
        # because they had a colonoscopy at age 50.
        {
            "routine_test_by_year": ["Colonoscopy"] + ["FIT"] * 25,
            "expected_colonoscopies": 1,
            "expected_fits": 16,
        },
        ... other 3 switching scenarios here
    ]
)
def test_switching_scenario(params, case):
    ...
```

Inside the test you'd build the step function and then do everything your current tests do. Except at the end you'd use the expected values from the case definition in your assertions.

Another advantage of this approach for me is that it puts the switching scenario right beside the expected numbers of colonoscopies and FITs, so it's easier for the reader to confirm that they agree with the expected counts.

Reply:
I restructured the scenarios as a single parametrized test `test_switching_scenario`. Much improved - thanks for the suggestion.